File: src/main/scalacode/com/huangning/scala/spark/sql/UDFandUDAF/UDAF.scala
package com.huangning.scala.spark.sql.UDFandUDAF

import org.apache.spark.sql.expressions.{MutableAggregationBuffer, UserDefinedAggregateFunction}
import org.apache.spark.sql.types.{DataType, DataTypes, IntegerType, StringType, StructType}
import org.apache.spark.sql.{DataFrame, Row, SparkSession}

/**
 * Author: 黄宁
 * Date: 2020/07/04
 * Target: 了解UDAF(UserDefinedAggregateFunction 用户自定义聚合函数 抽象类）
 * Note:
 * 1. 常用的聚合函数有count, avg, sum, min, max等等
 * 2. 特点：多对一
 * 3. 聚合函数必须和group by一起使用
 * 4. 和聚合函数一起出现的字段，必须要出新在group by的后面，比如：select name ,count(*) from table group by name
 */
class MyUDAF extends UserDefinedAggregateFunction{
  // 指定输入数据类型
  def inputSchema: StructType ={
    DataTypes.createStructType(Array(DataTypes.createStructField("uuuu", StringType, true)))
  }

  /**
   * 为每个分组的数据执行初始化
   * 两个部分初始化
   * 1. 在map端, 每个RDD分区内按照groupby的分组，每个分组都有个初始化的值
   * 2. 在reduce端给每个groupby分组的一个初始值
   */
  override def initialize(buffer: MutableAggregationBuffer): Unit = {
    buffer(0) = 0
  }

  // 每个组，有新的值进来的时候，进行分组对应的聚合值的计算 重要！！！
  override def update(buffer: MutableAggregationBuffer, input: Row): Unit = {
    buffer(0) = buffer.getAs[Int](0) + 1
  }

  /**
   * 最后merge的时候，在各个节点上的聚合值，要进行merge（也就是合并）
   * @param buffer1: reduce端给每个groupby分组的一个初始值
   * @param buffer2: 每个分区的每个groupby的结果
   */
  override def merge(buffer1: MutableAggregationBuffer, buffer2: Row): Unit = {
    buffer1(0) = buffer1.getAs[Int](0) + buffer2.getAs[Int](0)
  }

  // 聚合操作的时候，所处理的数据类型
  override def bufferSchema: StructType = {
    DataTypes.createStructType(Array(DataTypes.createStructField("QQQQ", IntegerType, true)))
  }

  // 最后返回一个最终的聚合值要和dataType的类型一一对应
  override def evaluate(buffer: Row): Any = {
    buffer.getAs[Int](0)
  }

  // 最终函数返回值的类型
  override def dataType: DataType = {
    DataTypes.IntegerType
  }

  // 多次运行 相同的输入总是相同的输出，确保一致性
  override def deterministic: Boolean = {
    true
  }
}

object UDAF {
  def main(args: Array[String]): Unit = {
    val spark: SparkSession = SparkSession.builder().appName("UDAF").master("local").getOrCreate()
    val nameList: List[String] = List[String]("zhangsan", "lisi", "wangwu", "zhaoliu", "zhangsan", "lisi", "zhangsan")

    // 将scala的List转成DataFrame
    import spark.implicits._
    val frame: DataFrame = nameList.toDF("name")
    frame.createOrReplaceTempView("students")

    // 注册UDAF函数
    spark.udf.register("NAMECOUNT", new MyUDAF())
    spark.sql("select name, NAMECOUNT(name) from students group by name").show(100)

  }
}

File: src/main/javacode/com/huangning/java/spark/sql/udf/UDAF.java
package com.huangning.java.spark.sql.udf;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.expressions.MutableAggregationBuffer;
import org.apache.spark.sql.expressions.UserDefinedAggregateFunction;
import org.apache.spark.sql.types.DataType;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class UDAF {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setMaster("local").setAppName("UDAF");
        JavaSparkContext sc = new JavaSparkContext(conf);
        SQLContext spark = new SQLContext(sc);

        // 通过动态Schema的方式将RDD转换成DataFrame
        JavaRDD<String> rdd = sc.parallelize(Arrays.asList("zhangsan", "lisi", "wangwu", "lisi", "wangwu"));
        JavaRDD<Row> rowRDD = rdd.map(new Function<String, Row>() {
            @Override
            public Row call(String s) throws Exception {
                // 将字符串分装成Row类型--用RowFactory
                return RowFactory.create(s);
            }
        });

        List<StructField> fields = new ArrayList<StructField>();
        fields.add(DataTypes.createStructField("name", DataTypes.StringType, true));
        StructType schema = DataTypes.createStructType(fields);
        Dataset<Row> df = spark.createDataFrame(rowRDD, schema);
        df.createOrReplaceTempView("students");

        /**
         * 注册一个UDAF函数，实现统计相同值的个数
         * 注意：这里可以自定义一个类继承UserDefinedAggregateFunction类也是可以的
         */
        spark.udf().register("NAMECOUNT", new UserDefinedAggregateFunction() {
            /**
             * 在RDD的每个分区的每个分组更新结果
             * @param buffer
             * @param input
             */
            @Override
            public void update(MutableAggregationBuffer buffer, Row input) {
                buffer.update(0, buffer.getInt(0) + 1);
            }

            /**
             * 合并update操作，可能是一个节点上发生的，也可能是多个节点上发生的
             * @param buffer1：大聚合的时候上一次聚合的结果
             * @param buffer2：新传入的update结果
             */
            @Override
            public void merge(MutableAggregationBuffer buffer1, Row buffer2) {
                buffer1.update(0, buffer1.getInt(0) + buffer2.getInt(0));
            }


            /**
             * 每个分区的每组初始化结果
             * @param buffer
             */
            @Override
            public void initialize(MutableAggregationBuffer buffer) {
                buffer.update(0,0);
            }

            /**
             * 指定输入字段的字段以及类型
             * @return
             */
            @Override
            public StructType inputSchema() {
                return DataTypes.createStructType(
                        Arrays.asList(
                                DataTypes.createStructField("name111", DataTypes.StringType, true)
                        )
                );
            }

            /**
             * 在进行聚合操作的时候，所需要处理数据的结果的类型
             */
            @Override
            public StructType bufferSchema() {
                return DataTypes.createStructType(Arrays.asList(DataTypes.createStructField("1111", DataTypes.IntegerType, true)));
            }

            /**
             * 最后翻一个和dataType方法类型要一直的类型，返回UDAF最后的计算结果
             * @param buffer
             * @return:
             */
            @Override
            public Object evaluate(Row buffer) {
                return buffer.getInt(0);
            }

            /**
             * 指定UDAF函数计算后返回的结果类型
             */
            @Override
            public DataType dataType() {
                return DataTypes.IntegerType;
            }


            /**
             * 确保一致性一般用true，用来标记针对给定的一组输入，UDAF是否总是生成相同的结果
             * @return
             */
            @Override
            public boolean deterministic() {
                return true;
            }
        });

        spark.sql("select name, NAMECOUNT(name) from students group by name").show(100);

        sc.stop();
    }
}
